issue #1326 projectStats [WIP]

dsych opened the pull request:
ProjectLimiter is undefined, which causes the problem.

humphd commented on src/bramble/client/main.js:
```diff
@@ -27,10 +27,11 @@ define([
     "bramble/client/StateManager",
-    "bramble/thirdparty/MessageChannel/message_channel"
-], function(Filer, ChannelUtils, EventEmitter, StateManager) {
+    "bramble/thirdparty/MessageChannel/message_channel",
```
Your order is wrong here, put `message_channel` last

dsych replied:
This fixes it. Thanks.

humphd commented on src/bramble/client/main.js:
```diff
@@ -136,8 +137,11 @@ define([
+
     // After calling Bramble.load(), Bramble.mount() specifies project root entry path info
-    Bramble.mount = function(root, filename) {
+    Bramble.mount = function(root, filename,projectLimits) {
```
Use the same name as how you require it in, which in line 34 is `ProjectLimiter`.

dsych replied:
projectLimits here is parameter that is passed from hosted.js that contains info about the maximum allowed project limits.

dsych commented:
For now, it seems to be working correctly. However, there is a warning:
>Deleted /7/projects/30/Untitled-1 Document still has 2 references. Did someone addRef() without listening for 'deleted'?
It pops up after i delete a new file. My guess is I am not cleaning-up the stats object in the checkLimits function.

humphd commented on src/bramble/client/ProjectLimiter.js:
```diff
@@ -0,0 +1,94 @@
+    //path has to be pointing to an existing file or to a file that has been just deleted. If not triggers incorrect behaviour.
+    ProjectLimiter.checkLimits = function (root,fs,oldSize){
+        var shell = new fs.Shell();
+        shell.ls(root, {recursive:true},function(err,entries){
```
If you're going to do it this way (i.e., async), which is fine, then your outer function needs to accept a `callback` function as one of the arguments, and you need to call that `callback` function when you're done doing the checks below.

humphd commented:
That error is from within Brackets, and it must mean that some code is still holding a reference to the doc after it gets removed.  It's not likely to be your bug, but I'll have to read your code in more detail later.

humphd commented on src/bramble/client/ProjectLimiter.js:
```diff
@@ -0,0 +1,94 @@
+    var ProjectLimiter = {};
+
+    var maxProjectState = {
+            fileSize: 0, // max size of any file in bytes. NOT USED now.
```
Be consistent with your indents, 4-spaces in Brackets.

humphd commented on src/bramble/client/ProjectLimiter.js:
```diff
@@ -0,0 +1,94 @@
+        projectSize: 0
+    };
+
+    ProjectLimiter.initMaxLimits =  function (projectState){
```
Usually the way you do this is to allow passing a so-called config object, and then you read off the values it has, doing some validation.  For example:

```js
ProjectLimiter.init = function(config) {
    _maxProjectState.fileSize = config.fileSize || 0; // set your default, I used 0
    _maxProjectState.fileCount = config.fileCount || 0; // etc
...
}
```

dsych replied:
When is this init function is supposed to be called?

humphd commented on src/bramble/client/ProjectLimiter.js:
```diff
@@ -0,0 +1,94 @@
+        maxProjectState = projectState;
+        console.log(projectState);
+    };
+   function _checkLimits(){
```
See my comment below about needing a callback.  What you'll want to do is return an object on your callback that has all this data, and leave it up to the caller what to do (maybe `console.log` or maybe use the values to do things):

```js
ProjectLimiter.checkLimits = function(root, fs, oldSize, callback) {
    ....
    // eventually you send back your data.  I'm hard coding it here, but you'd have values you calculate:
    callback(null, { fileSize: 6, fileCount: 25, ... });
```

dsych replied:
I see your point, will fix this part.

humphd commented on src/bramble/client/ProjectLimiter.js:
```diff
@@ -0,0 +1,94 @@
+    //accepts path be checked, filer and the oldSize of a file.
+    //path has to be pointing to an existing file or to a file that has been just deleted. If not triggers incorrect behaviour.
+    ProjectLimiter.checkLimits = function (root,fs,oldSize){
+        var shell = new fs.Shell();
```
Another way to do this would be to pass `root` and `fs` when you do `init()` so you can store those for the duration of the program--they won't change, so you likely don't want to have to pass them all the time.

dsych replied:
As I state in the comment before, root is not pathname to a root of the project, rather a constantly changing path to a file being modified. However, you are right to suggest storing the reference to a fs in the object state.

humphd commented on src/bramble/client/ProjectLimiter.js:
```diff
@@ -0,0 +1,94 @@
+    };
+
+    ProjectLimiter.newFile = function(){
+            currentProjectState.numOfFiles +=1;
```
It's interesting that for some of the data you're keeping track of it as it happens, and for other data you're walking the filetree to figure things out.  You should probably pick one.  I think I'd do the latter, since you're already walking the entire filetree, and could easily extract the file count that way.

dsych replied:
My naming of variables should have confused you. Root parameter in the checkLimits function is not actually a root, rather a pathname to a file we are working with right now. I never walk the whole file tree.
I see why it is confusing: I am passing a recursive option to the ls call, which is not really needed, since I only deal with the files anyways.

dsych commented:
@humphd please take a look at the comment above.

dsych commented:
@humphd I've addressed most of the problems you outlined. However, I am still not clear on the init function. Where do you want me to call it?

humphd commented on src/bramble/client/ProjectLimiter.js:
```diff
@@ -0,0 +1,69 @@
+
+    var _fs;
+
+    ProjectLimiter.initMaxLimits =  function (projectState,fs){
```
I'm starting to think that we should separate out the collection of stats from the concept of having limits.  We probably want to end up with 2 modules, one for `ProjectStats` and another for `ProjectLimits`, and the second would use the first to do its work.  I think that could happen in a follow-up PR if you want, and this one could just focus on the `ProjectStats` piece.

Having said that, let's evolve how we do the `init` method.  Since we've already wrapped the filesystem methods we care about, we now need to set our initial state by getting the project root dir that we'll be using for the duration of this instance (e.g., whatever is specified in `Bramble.mount(...)`) and walking that dir recursively, doing `fs.stat` on everything to build an initial picture of what we have.

This step is important because there are different cases that we might encounter on startup:
* The user is starting a brand new project, and there are no files in the filesystem
* The user is opening an existing project, and the filesystem is full of files.

The wrapped filesystem methods, and this code should be reference an internal object that caches various pieces of info (e.g., file size).  After we do this initial walk of the root dir, the wrapped fs methods will capture any changes that happen later on.

humphd replied:
NOTE: we'll want to have `init` take a `callback` function, since this is going to need to happen asynchronously.

humphd commented on src/bramble/client/ProjectLimiter.js:
```diff
@@ -0,0 +1,69 @@
+        console.log(projectState);
+    };
+
+    ProjectLimiter.newFile = function(){
```
We can remove this now, since we'll do it in wrapped function calls.

humphd commented on src/bramble/client/ProjectLimiter.js:
```diff
@@ -0,0 +1,69 @@
+
+    //accepts path be checked,the oldSize of a file, and an optional callback.
+    //path has to be pointing to an existing file or to a file that has been just deleted. If not triggers incorrect behaviour.
+    ProjectLimiter.checkLimits = function (pathname,oldSize,callback){
```
Because we'll do a mix of walking the root dir at startup, and also wrapping all modifying fs calls after that, we can make methods that return data synchronous: we'll just be looking at values in our cache, or calculating things based on it.

Probably we want to split this method into a bunch of smaller, more specific stats.  Of the top of my head:

```js
ProjectStats.getTotalSize = function(){...} // returns bytes of files in the root
...
ProjectStats.getLargestFileSize = function(type) {
  if(type) {
    // get largest file size for the given type (e.g., image, html, ...).
    // We could do this by extension (.html) or maybe by name (image)
  } else {
    // get larges file size for any type of file (e.g., ignore type)
  }
};
...
ProjectStats.getFileCount = function(type) {
  if(type) {
    // get file count for the given type (e.g., image, html, ...).
    // We could do this by extension (.html) or maybe by name (image)
  } else {
    // get file count for any type of file (e.g., ignore type)
  }
};
...
ProjectStats.hasIndexFile = function() {
  // returns true/false depending on whether root/index.html exists
  // we would love to know this when we publish, so we can warn users.
};
...
ProjectStats.usesCSS = function() {
  // returns true/false depending on whether the project has CSS files
};
...
ProjectStats.usesJS = function() {
  // returns true/false depending on whether the project has JS files
};
...
etc
```

humphd commented on src/bramble/client/ProjectLimiter.js:
```diff
@@ -0,0 +1,69 @@
+                    _currentProjectState.projectSize -= oldSize;
+                    //_checkLimits();
+                }
+                else{
```
`} else {` is the preferred way in JS.

humphd commented on src/bramble/client/ProjectLimiter.js:
```diff
@@ -0,0 +1,69 @@
+                    size: 0
+                };
+                //deleted from file
+                if(stats.size < oldSize){
```
Probably all this logic with old vs. new file sizes can go away with the approach described above.  We can just record whatever the existing size is in our cache.

humphd commented on src/bramble/client/ProjectLimiter.js:
```diff
@@ -0,0 +1,69 @@
+        });
+    };
+    return ProjectLimiter;
+});
```
Add a new line at the end of this file.

humphd commented on src/bramble/client/main.js:
```diff
@@ -27,10 +27,11 @@ define([
+    "bramble/client/ProjectLimiter",
     "bramble/thirdparty/MessageChannel/message_channel"
-], function(Filer, ChannelUtils, EventEmitter, StateManager) {
+], function(Filer, ChannelUtils, EventEmitter,StateManager,ProjectLimiter) {
```
`, ProjectStats)` space after comma please.

humphd commented on src/bramble/client/main.js:
```diff
@@ -506,11 +510,19 @@ define([
             // With successful fs operations that create, update, delete, or rename files,
             // we also trigger events on the bramble instance.
-            function genericFileEventFn(type, filename, callback) {
+            function genericFileEventFn(type, filename,callback,oldSize) {
```
We can undo all these changes, since we'll do it internally on the `_fs` ourselves.

dsych commented:
@humphd Thanks a lot for your feedback, I'll be working on it throughout this week.

dsych commented:
@humphd Please review the implementation of ProjectStats as you suggested.

humphd commented on src/bramble/client/main.js:
```diff
@@ -147,6 +148,7 @@ define([
             return;
         }
 
+        ProjectStats.init(root,function(){});
```
You're going to have a race condition here, since you are calling `init` and it will take time to walk the root dir, but then you immediately tell the instance to `mount` below.  You are also silently ignoring the error case.  You need to have them happen in succession.  Do it like this:

```js
ProjectsStats.init(root, function(err) {
  if(err) {
    // Deal with error case here, probably something like:
    setReadyState(Bramble.ERROR, new Error("Unable to access filesystem: ", err));
    return;
  }
  _instance.mount(root, filename);
});
```

humphd commented on src/document/DocumentCommandHandlers.js:
```diff
@@ -123,9 +123,10 @@ define(function (require, exports, module) {
 
     /**
      * index to use for next, new Untitled document
-     * @type {number}
```
Rebase to deal with this.

humphd commented on src/bramble/client/ProjectStats.js:
```diff
@@ -0,0 +1,154 @@
+            }
+
+            _fs.stat(path,function(err,stats){
+                _projectInfo.size += stats.size;
```
Instead of doing this, just keep track of paths and sizes:

```js
_projectInfo[path] = stats.size;
```

humphd commented on src/bramble/client/ProjectStats.js:
```diff
@@ -0,0 +1,154 @@
+                return next();
+            }
+
+            _fs.stat(path,function(err,stats){
```
You always need to deal with the `err` case in your callback.  Also, you should deal with dir vs. file nodes, since you only really care about files.  You *could* keep track of both, and just record `0` for the size of a dir?

dsych commented:
@humphd I have implemented new features as advised. Please review when you have time.

dsych commented:
I was considering  addition of ` _extFromMime(mime)` function to src\filesystem\impls\filer\lib\content.js . However, I was not sure if you wanted it there, so for now it is in my module.

humphd commented on locales/el/editor.properties:
```diff
@@ -105,8 +105,12 @@ CMD_TOGGLE_QUICK_DOCS=Γρήγορα έγγραφα
 DND_UNSUPPORTED_FILE_TYPE=μη υποστηριζόμενος τύπος αρχείου
```
Before I review this again, you need to fix these changes that aren't related to your patch.  Let me know when you've done all that.

dsych replied:
@humphd did the rebase.

Pomax commented on src/bramble/client/ProjectStats.js:
```diff
@@ -1,154 +1,259 @@
     "use strict";
 
-    function ProjectStats() {};
+    var ProjectStats = {};
```
This doesn't seem right - this is an Object definition, we can't just rewrite that to a singleton object.

humphd commented on src/bramble/client/ProjectStats.js:
```diff
@@ -0,0 +1,258 @@
+    // original writeFile
+    var _innerWriteFile = _fs.writeFile;
+    // overwrite original writeFile and add bookkeeping of project state and call original writeFile.
+    _fs.writeFile = function(filename, data, options, callback){
```
Let's simplify what you're doing here a bit, and just normalize the args before we use them (untested code follows):

```js
_fs.writeFile = function(filename, data, options, callback){
    if(typeof options === "function") {
        callback = options;
        options = {};
    }

    var size = data.length;

    _innerWriteFile.call(_fs, filename, data, options, function(err){
        if(err) {
            return callback(err);
        }
        // overwrite or add a new size for a given filename
        _cache[filename] = size;
        callback();
    });
};
```

humphd commented on src/bramble/client/ProjectStats.js:
```diff
@@ -0,0 +1,258 @@
+    _fs.writeFile = function(filename, data, options, callback){
+        function update(){
+            //only check once _root is established
+            if(_root){
```
Since you blow-away the `_cache` when we call `init` we don't really need to do this check.

humphd commented on src/bramble/client/ProjectStats.js:
```diff
@@ -0,0 +1,258 @@
+    ProjectStats.hasIndexFile = function() {
+        // Loop through all paths, return true if one of them is root/index.html
+        return !!(Object.keys(_cache)).find(function(path) {
+            return path === _Path.join(_root, "index.html");
```
One optimization we can do here is to cache the full path to the index file vs. creating it every time we go through the loop:

```js
ProjectStats.hasIndexFile = function() {
         // Loop through all paths, return true if one of them is root/index.html
         var indexPath = Path.join(_root, "index.html");

         return !!(Object.keys(_cache)).find(function(path) {
             return path === indexPath;
...
```

humphd commented on src/bramble/client/ProjectStats.js:
```diff
@@ -0,0 +1,258 @@
+    // returns bytes of files in the root   
+    ProjectStats.getTotalProjectSize = function() {
+        // Sum up all sizes in the cache and return
+        return _root ? (Object.values(_cache)).reduce(function(acc, val) {
```
Simplify this a bit so it reads more cleanly:

```js
if (!_root) {
    return 0;
}

return (Object.values...)
```

humphd commented on src/bramble/client/ProjectStats.js:
```diff
@@ -0,0 +1,258 @@
+
+    // returns project file count
+    ProjectStats.getFileCount = function (){
+        return _root ? Object.keys(_cache).length : 0;
```
Same thing here, break this out into a separate `if`

humphd commented on src/bramble/client/ProjectStats.js:
```diff
@@ -0,0 +1,258 @@
+
+    // function converts mimes into corresponding file formats.
+    // mime format e.g. "text/html". Also supports "text/*", which returs an array of all corresponding types.
+    function _extFromMime(mime) {
```
You asked about whether to include the `content` module and for sure you should vs. repeating this code.  It makes it harder to maintain, because we end up with 2 versions of the same thing.  Just include the module at the top in your `define` call.  If you're not sure what I mean, let me know.

humphd replied:
Ignore this, let's just erase this for now.

humphd commented on src/bramble/client/ProjectStats.js:
```diff
@@ -0,0 +1,258 @@
+
+    // returns an array of filenames for specified extension
+    // adds period if not provided
+    ProjectStats.usesExtension = function(fileExtensions, callback) {
```
Let's drop this for now, I don't know exactly how we'll use it, so I don't think we should include yet.

dsych commented on src/bramble/client/ProjectStats.js:
```diff
@@ -0,0 +1,258 @@
+
+    var _shell = new _fs.Shell();
+    var _Path = Filer.Path;
+    var _root;
```
`var _shell = new _fs.Shell();` on line 17 is supposed to go after the _fs declaration or it will give an error.

humphd replied:
Right, just move it up above but below that.

dsych commented:
@humphd please review changes on your convenience.

humphd commented on src/bramble/client/ProjectStats.js:
```diff
@@ -0,0 +1,135 @@
+            if(err) {
+               return callback(err);
+            }
+            callback(null);
```
You can just do `callback()` and it will do the same thing.

humphd commented on src/bramble/client/ProjectStats.js:
```diff
@@ -0,0 +1,135 @@
+    ProjectStats.getTotalProjectSize = function() {
+        // Sum up all sizes in the cache and return
+        if(_root){
+            (Object.values(_cache)).reduce(function(acc, val) {
```
Missing a `return` here.

humphd commented on src/bramble/client/ProjectStats.js:
```diff
@@ -0,0 +1,135 @@
+    // returns project file count
+    ProjectStats.getFileCount = function (){
+        if(_root){
+            Object.keys(_cache).length;
```
Same thing here, you need to `return`

humphd commented on src/bramble/client/main.js:
```diff
@@ -147,7 +149,15 @@ define([
         }
 
-        _instance.mount(root, filename);
+        ProjectStats.init(root,function(err){
```
Space after `,`

dsych commented:
@humphd pushed some more fixes as requested. 
I have an idea of how we can use `usesExtension()` function. When the project will be ready to implement projectLimiter concept, the method could be used to identify biggest media format files and give suggestions on where to host them. Maybe even use some external API to just select a file and then transfer it to a remote. I know this is really far-fetched due to the current phase of the project, however I could work on something like that for my next releases. Let me know what you think.

dsych commented:
@humphd are there any updates on the review?

humphd commented on src/bramble/client/ProjectStats.js:
```diff
@@ -0,0 +1,141 @@
+],function(Filer){
+    "use strict";
+
+    function ProjectStats() {};
```
This doesn't need to be a function, it should be an object that you extend with methods below:

```js
var ProjectStats = {};
```

humphd commented on src/bramble/client/ProjectStats.js:
```diff
@@ -0,0 +1,141 @@
+    // walk the whole file tree and record each path and file size
+    ProjectStats.init = function(root, callback){
+        function addSize(path, next){
+            // If the current path is a directory, record 0 for size
```
This comment seems out of date.  Remove it.

humphd commented on src/bramble/client/ProjectStats.js:
```diff
@@ -0,0 +1,141 @@
+        _cache = {};
+
+        // walk the root
+        _shell.find(_root, { exec:addSize }, function(err){
```
Because you aren't doing anything specific with the final callback, except checking `err`, it's common to not do anything, and just use the callback passed to you instead:

```js
_shell.find(_root, { exec:addSize }, callback);
```

If you were doing something (e.g., using an object returned in the callback, changing a variable, etc.) they you'd do what you did here.

humphd commented on src/bramble/client/ProjectStats.js:
```diff
@@ -0,0 +1,141 @@
+
+    };
+
+    // original unlink
```
Let's do all of this only after `init` is called.  Move lines 61-110 into `init` above.

humphd commented on src/bramble/client/ProjectStats.js:
```diff
@@ -0,0 +1,141 @@
+                delete _cache[oldPath];
+            }
+            
+            // call callback regardless
```
Remove this, it doesn't communicate anything more than the code does.

humphd commented on src/bramble/client/ProjectStats.js:
```diff
@@ -0,0 +1,141 @@
+                delete _cache[pathname];
+            }
+            
+            // call callback regardless
```
Remove this, it doesn't communicate anything new that the code doesn't already.

humphd commented on src/bramble/client/ProjectStats.js:
```diff
@@ -0,0 +1,141 @@
+
+    // returns true/false depending on whether root/index.html exists
+    ProjectStats.hasIndexFile = function() {
+        var index = _Path.join(_root, "index.html");
```
This needs `_root` to exist, so we should deal with the case of it not being set yet:

```js
if(!_root) {
  return false;
}
...
```

humphd commented on src/bramble/client/ProjectStats.js:
```diff
@@ -0,0 +1,141 @@
+        // Sum up all sizes in the cache and return
+        if(_root){
+            return (Object.values(_cache)).reduce(function(acc, val) {
+                return acc + (val ? val : 0);
```
Notice how much this code is getting indented?  A good pattern to follow is that you deal with error cases or edge cases first, and return early.  Then you can do your main code and not indent it deep in if statements:

```js
if(!_root) {
  return 0;
}

return (Object.values...
```

humphd commented on src/bramble/client/ProjectStats.js:
```diff
@@ -0,0 +1,141 @@
+    // returns project file count
+    ProjectStats.getFileCount = function (){
+        if(_root){
+            return Object.keys(_cache).length;
```
Same thing here, deal with `!_root` first and return early, then do your main code without being in an if.

humphd commented on src/bramble/client/main.js:
```diff
@@ -27,10 +27,11 @@ define([
     "thirdparty/filer/dist/filer.min",
     "bramble/ChannelUtils",
-    "bramble/thirdparty/EventEmitter/EventEmitter.min",
+    "bramble/thirdparty/EventEmitter/EventEmitter",
```
Still not dealt with.  Please don't ask for a review while you have outstanding issues from a previous review, it wastes everyone's time.

humphd commented on src/bramble/client/main.js:
```diff
@@ -110,13 +111,30 @@ define([
+    };
+
+    // Return true/false based on the existance of the index.html
+    Bramble.hasIndexFile =  function(){
```
Remove extra space after `=`

humphd commented on src/bramble/client/main.js:
```diff
@@ -147,7 +165,15 @@ define([
-        _instance.mount(root, filename);
+        ProjectStats.init(root, function(err){
+            if(err) {
+                // Deal with error case here, probably something like:
```
Get rid of this comment.

dsych commented:
@humphd I apologize, if I came out rude. It's just you've been so quick with your reviews before, I assumed there was a problem in this time.

humphd commented:
Thanks for saying that, and it's totally OK.  I know you're eager to keep going with this, which is great!  My problem is only that I have so many eager students all wanting reviews at the same time.  We'll get there.

dsych commented:
@humphd  just added some new changes, please review. 
Looking back at all of the styling problems like missing a space in function arguments or having a curly brace ( "{" || "}" ) in a wrong place, I looked at the current version of the .eslintrc.json file; it does not define any rules for most of the styling guidelines you enforce. On the other hand, should we define all of the rules properly in the esling file, we can then use a --fix option in order to avoid wasting time on fixing small styling errors. So the question is: "Should I file an issue for fixing this or is there anything that I am not aware of that makes said feature implementation impossible?".

humphd commented on src/bramble/client/ProjectStats.js:
```diff
@@ -0,0 +1,137 @@
+    };
+    // NOTE: THIS WILL DESTROY DATA! For error cases only, or to wipe the disk.
+    ProjectStats.formatFileSystem = function(flags, callback){
+        _fs = new Filer.FileSystem(flags, callback);
```
Because you create a new `_fs` you will also need to recreate `_shell` here too.

dsych replied:
@humphd I am not really sure if that is required based of the original code that avoids doing it. [Here](https://github.com/mozilla/brackets/blob/master/src/bramble/client/main.js#L118)

humphd replied:
The reason we don't need it there, but do need it here, is that you are holding a long-lived instance of this shell, and then using it later.  If you swap out the filesystem that's bound to the shell, it will break later on.  You could stop holding this reference, and just creating it whenever you need it.

dsych replied:
Sorry, I got a little confused with main.js.

humphd commented on src/bramble/client/ProjectStats.js:
```diff
@@ -0,0 +1,137 @@
+                _cache[path] = stats.size;
+                next();                
+            });
+        
```
Remove this blank line.

humphd commented on src/bramble/client/ProjectStats.js:
```diff
@@ -0,0 +1,137 @@
+
+        // walk the root
+        _shell.find(_root, { exec:addSize }, callback);
+
```
Remove this blank line.

humphd commented on src/bramble/client/ProjectStats.js:
```diff
@@ -0,0 +1,137 @@
+        _cache = {};
+
+        // walk the root
+        _shell.find(_root, { exec:addSize }, callback);
```
Move this to the bottom of this function at line 133, so we don't have a timing issue.

humphd commented on src/bramble/client/main.js:
```diff
@@ -110,13 +111,30 @@ define([
+    };
+
+    // Return total project size
+    Bramble.getTotalProjectSize = function(){
```
These functions shouldn't go on `Bramble` but on `bramble` (i.e., on the instance).  The reason is that they are closely bound to the project root, and `bramble` is the instance created when you call `Bramble.mount()` on a project root dir.  Put it on `_instance` or in the constructor for the `BrambleProxy` (e.g., on `self`).

dsych commented:
@humphd pushed some changes, however I am not sure on one thing, please review the comment above.

dsych commented:
@gideonthomas @humphd I added some linting rules just to make review less painless, they could be removed later if you wish. Moreover, I implemented the usesExtension function that we discussed today during class. For now, _extFromMime is inside projectState, later when this code is ready to land, I intend to move it to src\filesystem\impls\filer\lib\content.js

humphd commented on src/bramble/client/ProjectStats.js:
```diff
@@ -0,0 +1,263 @@
+/* eslint
```
Brackets already has an eslint config, see https://github.com/mozilla/brackets/blob/master/.eslintrc.json.  Why not just have this code use that?  My guess is that we aren't passing this file to eslint in the Gruntfile.

dsych replied:
It goes back to the fact that you are not doing any style checking with you eslint configurations

humphd replied:
I hear you.  I think it's really dangerous for us to start doing manual style changes within Brackets, because it will totally break our ability to merge with Adobe's upstream.  Any style changes we do have to be done automatically by a tool, for example, prettier.

If we want to do this in Brackets, my suggestion is as follows:
* Create a branch `adobe-upstream` which tracks Adobe's `master` branch
* Create a branch `adobe-upstream-pretty` which is `adobe-upstream` with auto-formatting applied.
* Figure out some kind of machinery to automatically keep the two above branches always up-to-date (maybe with git, maybe with travis, maybe with some git hook)
* Create some kind of grunt/npm task/git pre-commit hook/etc that runs our autoformatting on all code we land on `master`
* Do all of our future merges of `master` against `adobe-upstream-pretty`.  We don't do this very often.

If you're interested in doing this work, we can file a new bug.  I don't want to introduce separate lint rules per file. cc @Pomax

dsych replied:
@humphd I see your point, that why earlier I said:

>  I added some linting rules just to make review less painless, they could be removed later if you wish

The point of this linting is not to introduce my own new rules to separate files. On the contrary, my goal is to reduce review time and effort that is needed from you. The rules in my file are based on suggestions you gave me throughout this review.

dsych replied:
In addition, I am not sure if you've actually reached consensus [here](https://github.com/mozilla/thimble.mozilla.org/issues/1800) about which approach to take.

humphd replied:
I don't think we have 100% consensus yet, but I want to see us try something anyway.  The least controversial one to do imho is the JS formatting.  Let's just do it.

Also, please remove this eslint comment, we can't do this in this file for reasons we've discussed.

humphd commented on src/bramble/client/ProjectStats.js:
```diff
@@ -0,0 +1,263 @@
+        _shell.find(_root, { exec:addSize }, callback);
+    };
+
+    // accepts an array of mimes and/or file extensions
```
I don't want to hold this PR up any longer, and adding new things to it is a bad idea (we've already taken way too long).  It's better to do small, incremental changes to a piece of code vs. trying to ram through a lot of things in one go.  To you, this is all related work, and should probably happen at the same time.  However, it's more efficient if we break the work up into separate PRs that build on each other.

Please remove this and start a new PR for this extra work.

dsych commented:
@humphd I did as you requested, will open a new PR for subsequent changes.

humphd commented on src/bramble/client/ProjectStats.js:
```diff
@@ -0,0 +1,132 @@
+    "use strict";
+
+    var ProjectStats = {};
+    var _Path = Filer.Path;
```
Change this to `Path` vs. `_Path` please.

humphd commented on src/bramble/client/ProjectStats.js:
```diff
@@ -0,0 +1,132 @@
+                options = {};
+            }
+
+            _innerWriteFile.call(_fs, filename, data, function(err){
```
We forgot to include the possible [`options` arg](https://github.com/filerjs/filer#writeFile) here, should be:

```js
_innerWriteFile.call(_fs, filename, data, options, function(err) {
```

humphd commented:
@gideonthomas if you want to review this, please do so soon.  I want to finish this.

dsych commented:
@humphd please review, hopefully this is it. Let me know if I need to rebase it.

humphd commented:
Somehow we have a whole bunch of commits in here that don't belong to this patch.  I blogged about how to fix this kind of thing at http://blog.humphd.org/git-does-what-you-tell-it/.  @dsych can you remove anything that doesn't belong in this PR?

dsych commented:
@humphd hopefully, this is what you desired.

gideonthomas commented:
@humphd I think I'm going to say let's go with this fix and iterate on this (based on how we will use it). Requesting one final quick review and then we can merge this.

dsych commented:
@humphd @gideonthomas I was working on related bug that required me to log-in into thimble. Before the log-in, filesystem behaved properly, however after login, it fails to properly do any updates and throws on every change. Here is the error:

> callback is not a function

I examined the call stack and it seems that is being thrown at callback call in every respective function. E.g. https://github.com/mozilla/brackets/pull/620/files?diff=unified#diff-51ad1fd8eb0c794c6deea6808133b61bR53
I believe, we need to perform similar check as we did in [write](https://github.com/mozilla/brackets/pull/620/files?diff=unified#diff-51ad1fd8eb0c794c6deea6808133b61bR61) before calling the callback. I am not sure what might be causing those functions to be called without callback.

humphd commented:
Do I need to revert this?  Is it breaking on master now?  If so, I'll undo this merge, and you can fix then do a new PR.

Let me know ASAP.  We can't leave master broken.

dsych commented:
@humphd revert it for now. I will verify what is causing it. In the meantime,are you aware of any changes that happen to how that happen when someone logs in?

humphd commented:
I've reverted in https://github.com/mozilla/brackets/commit/5158dfb948088767219ba88d39b3295b14bb9264.  @dsych thanks for catching this so quickly.  This kind of thing happens to all of us, so don't feel badly.  What's great is that you did QA as soon as it landed, and let us know there was an issue.

Fix it in a new PR and we'll land that when it's ready.

dsych commented:
@humphd just tried it with the clean master without any changes, which produces the same error... Can you try the whole process on your computer, please? Maybe it is just me.

humphd commented:
I've got `master` of Brackets and Thimble running locally, and I'm able to login, add files, delete files, no errors.

dsych commented:
@humphd This might be false alarm, can I ask you to perform one more check? Can you repeat the same process, but using my branch for brackets, please?

dsych commented:
@humphd I apologize for the hustle. It was a problem with my WM after all. Just tried using my code with thimble and everything works perfectly. 
Again, sorry for the disturbance, I only had best intentions in mind.

humphd commented:
So, to confirm, should I revert-my-revert commit and reapply this?  Please test everything thoroughly before saying 'yes' so we don't have to do this again.

dsych commented:
@humphd Just finished testing with multiple file type before and after log-in. Everything seems to be working properly.

humphd commented:
OK, I've re-applied this patch, and it's building on master https://travis-ci.org/mozilla/brackets/builds/221747036.  When that finishes, please test a lot on https://bramble.mofostaging.net/en-US/ and if you spot something, let me know.

dsych commented:
@humphd multiple file types were successfully modified, renamed, deleted and added. I think we are good.
